### PR TITLE
Provide, Inject and more hooks for component lifecycle

### PR DESCRIPTION
**Pull request by reandimo**
Hey @samsonasik ,

This is my solution to add most of the lifecycle hooks according to Vue.js documentation [here ](https://vuejs.org/api/options-lifecycle.html) and also add the abilityu to provide and inject data, useful for nesting components.

How to use:

Let's take for example `portfolio.js`
```javascript
import createPage from './create-page.js';

let portfolio = createPage(
    'portfolio',
    {
        portfolio : [
            {
                title : 'loading...',
                image : '',
                link  : '',
            }
        ]
    },
    // inject
    ['passUser'],
    // computed
    {
        user(){
            return this.passUser();
        }
    },
    // provide
    null,
    {
        search: function (e = null) {
            let keyword = e !== null
                ? e.target.value
                : '';

            if (typeof this.$store.state.portfolio.portfolio[keyword] !== 'undefined') {
                this.portfolio = this.$store.state.portfolio.portfolio[keyword];

                return;
            }

            if (sessionStorage.getItem('search-portfolio-' + keyword)) {
                let portfolio  = JSON.parse(sessionStorage.getItem('search-portfolio-' + keyword));
                this.$store.commit('portfolio/search', { keyword: keyword, value: portfolio });
                this.portfolio = portfolio;

                return;
            }

            (async () => {
                await new Promise( (resolve) => {
                    fetch(
                        '/api/portfolio?keyword=' + keyword,
                        {
                            method: 'GET',
                            headers: {
                                'Accept': 'application/json',
                            }
                        }
                    ).then(response =>  resolve(response.json()));
                }).then(result => this.portfolio = result);

                this.$store.commit('portfolio/search', { keyword: keyword, value: this.portfolio });
            })();
        }
    },
    // hooks
    {
        updated: function() {
            this.$nextTick(() => this.search());
        },
        activated: function(){
            console.log(this.user)
            console.log('Activated!!!!')
        },
        deactivated: function(){
            console.log('Deactivated!!!!')
        }
    } 
);

export default portfolio;
```

Note that now I can use more hooks and also inject the function `passUser` and user the computed variable `user` provided from the parent `app.js`:

```javascript
import home       from './home.js';
import about      from './about.js';
import contact    from './contact.js';
import portfolio  from './portfolio.js';
import store      from './store.js';
import createPage from './create-page.js';

const { createRouter, createWebHistory } = VueRouter;

const routes = [
    { path: '/', component: home, meta: { title: 'Home'}  },
    { path: '/about', component: about, meta: { title: 'About'}  },
    { path: '/contact', component: contact, meta: { title: 'Contact'}  },
    { path: '/portfolio', component: portfolio, meta: { title: 'Portfolio'}  },
    { path: '/:pathMatch(.*)*', component: createPage('notFound'), meta: { title: 'Page Not Found' } }
];

const router = createRouter({
    history: createWebHistory(),
    routes: routes,
    base: '/',
    linkExactActiveClass: "active"
});

router.afterEach(to => document.title = to.meta.title);

const app = Vue.createApp({
    data() {
        return {
            user: {
                id: 1,
                username: 'test this thing',
                email: 'email@email.com',
            }
        }
    },
    methods: {
    },
    provide() {
        return {
            passUser: () => this.user,
        }
    }
});
app.use(router);
app.use(store);

router.isReady().then(() => app.mount('#root'))
```

Let me know what you think.

Best Regards,
Renan

### CODE REVIEW

File: public/js/create-page.js [hunk 1]
```diff
@@ -1,13 +1,16 @@
 const { h, compile } = Vue;
 
-const createPage = (name, data = {}, methods = {}, updated = () => {}) => Vue.defineComponent({
+const createPage = (name, data = {}, inject = [], computed = {}, provide = null, methods = {}, hooks = {}) => Vue.defineComponent({
```
Comment (samsonasik) [inline public/js/create-page.js:3]: I didn't check, but new params should be after last one as can possibly break existing arg call used as "updated" param fill:

https://github.com/samsonasik/ci4-vue/blob/24aa7f78e95a0b2f0022e27264aaa44dcf9128a2/public/js/portfolio.js#L51

Comment (reandimo) [reply public/js/create-page.js:3]: My bad I didn't update the protfolio.js. Will do in a moment!

**Comment by reandimo** [comment]
@samsonasik portfolio.js updated my friend, feel free to check!

File: public/js/portfolio.js [hunk 2]
```diff
@@ -11,6 +11,9 @@ let portfolio = createPage(
             }
         ]
     },
+    [],
+    {},
+    null,
```
Comment (samsonasik) [inline public/js/portfolio.js:16]: Can these new arg moved to last with default value so no need to fill when not needed?

Comment (reandimo) [reply public/js/portfolio.js:16]: Sure, will fix it in a moment

Comment (reandimo) [reply public/js/portfolio.js]: Done, you can check now.

File: public/js/portfolio.js [hunk 3]
```diff
@@ -48,9 +48,20 @@ let portfolio = createPage(
...
+        activated: function(){
+            console.log('Activated!')
+        },
+        deactivated: function(){
+            console.log('Deactivated!')
+        }
+    },
+    [],
+    {},
+    null,
```
Comment (samsonasik) [inline public/js/portfolio.js:64]: Could this change on portfolio.js be reverted and keep working?

Comment (reandimo) [reply public/js/portfolio.js]: Yep, I just removed those from the template on protfolio.js.

File: public/js/portfolio.js [hunk 4]
```diff
@@ -48,9 +48,17 @@ let portfolio = createPage(
...
+    {
+        updated: function() {
+            this.$nextTick(() => this.search());
+        },
+        activated: function(){
+            console.log('Activated!')
+        },
+        deactivated: function(){
+            console.log('Deactivated!')
+        }
```
Comment (samsonasik) [inline public/js/portfolio.js:60]: these part can be removed.

Comment (reandimo) [reply public/js/portfolio.js]: Done bro, check now

**Review by samsonasik — APPROVED** [review]
Looks good, thank you 👍

**Pull request merged**